File: Code/Chapter 13/C#/SqlServerAggregate/Concatenator.cs
using System;
using System.Data;
using System.Data.Sql;
using System.Data.SqlTypes;
using Microsoft.SqlServer.Server;

using System.Text;

namespace SqlServerAggregate
{
   [Serializable]
   [Microsoft.SqlServer.Server.SqlUserDefinedAggregate(Format.UserDefined, 
      MaxByteSize=8000)]
   public struct Concatenator : IBinarySerialize
   {
      private StringBuilder sb;
      public void Init()
      {
         sb = new StringBuilder();
      }

      public void Accumulate(SqlString Value)
      {
         sb.Append(Value);
         sb.Append(",");
      }

      public void Merge(Concatenator Group)
      {
         Accumulate(Group.ToString());
      }

      public override string ToString()
      {
         return sb.ToString();
      }

      public SqlString Terminate()
      {
         sb.Remove(sb.Length - 1, 1);
         return sb.ToString();
      }

      #region IBinarySerialize Members

      public void Read(System.IO.BinaryReader r)
      {
         sb = new StringBuilder();
         sb.Append(r.ReadString());
      }

      public void Write(System.IO.BinaryWriter w)
      {
         if (sb.Length > 0)
            w.Write(sb.ToString());
      }

      #endregion
   }
}

//Create Table Person
//(
//   PersonID int identity primary key,
//   PersonName varchar(200),
//   PersonRole varchar(200)
//)
//GO

//Insert into Person (PersonName, PersonRole) Values ('Sahil Malik', 'Author')
//Insert into Person (PersonName, PersonRole) Values ('Erick Sgarbi', 'Reviewer')
//Insert into Person (PersonName, PersonRole) Values ('Frans Bouma', 'Reviewer')
//Insert into Person (PersonName, PersonRole) Values ('Jonathan Hassell', 'Lead Editor')
//GO


//Create Assembly SqlServerAggregate
//from
//'C:\PlayArea\SqlServerUDT\bin\Debug\SqlServerAggregate.dll'
//GO

//CREATE AGGREGATE Concatenator( @instr nvarchar(400) )
//RETURNS nvarchar(MAX)
//EXTERNAL NAME [SqlServerAggregate].[SqlServerAggregate.Concatenator]
//GO

//Select PersonRole, dbo.Concatenator(PersonName) from Person where PersonRole = 'Reviewer' Group By PersonRole

File: Code/Chapter 13/VB.NET/SqlServerAggregate/Concatenator.vb
Imports System
Imports System.Data
Imports System.Data.Sql
Imports System.Data.SqlTypes
Imports Microsoft.SqlServer.Server
Imports System.Text

<Serializable()> _
<Microsoft.SqlServer.Server.SqlUserDefinedAggregate(Format.UserDefined, _
  MaxByteSize:=8000)> _
Public Structure Concatenator
   Implements IBinarySerialize

   Private sb As StringBuilder

   Public Sub Init()
      sb = New StringBuilder
   End Sub

   Public Sub Accumulate(ByVal value As SqlString)
      sb.Append(value)
      sb.Append(",")
   End Sub

   Public Sub Merge(ByVal value As Concatenator)
      Accumulate(value.ToString())
   End Sub

   Public Overrides Function ToString() As String
      Return sb.ToString()
   End Function

   Public Function Terminate() As SqlString
      sb.Remove(sb.Length - 1, 1)
      Return sb.ToString()
   End Function

#Region "IBinarySerialize Members"
   Public Sub Read(ByVal r As System.IO.BinaryReader) _
      Implements IBinarySerialize.Read
      sb = New StringBuilder()
      sb.Append(r.ReadString())
   End Sub

   Public Sub Write(ByVal w As System.IO.BinaryWriter) _
      Implements IBinarySerialize.Write
      If (sb.Length > 0) Then
         w.Write(sb.ToString())
      End If
   End Sub
#End Region
End Structure

